File: src/main/java/com/github/jactor/persistence/service/UserService.java
package com.github.jactor.persistence.service;

import com.github.jactor.persistence.dto.UserDto;
import com.github.jactor.persistence.entity.UserEntity;
import com.github.jactor.persistence.repository.UserRepository;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class UserService {

  private final UserRepository userRepository;

  @Autowired
  public UserService(UserRepository userRepository) {
    this.userRepository = userRepository;
  }

  public Optional<UserDto> find(String username) {
    return userRepository.findByUsername(username).map(UserEntity::asDto);
  }

  public Optional<UserDto> find(Long id) {
    return userRepository.findById(id).map(UserEntity::asDto);
  }

  public UserDto saveOrUpdate(UserDto userDto) {
    UserEntity userEntity = new UserEntity(userDto);
    userRepository.save(userEntity);

    return userEntity.asDto();
  }

  public List<String> findUsernamesOnActiveUsers() {
    return userRepository.findByUserType(UserEntity.UserType.ACTIVE).stream()
        .map(UserEntity::getUsername)
        .collect(Collectors.toList());
  }
}

File: src/main/kotlin/com/github/jactor/persistence/service/UserService.kt
package com.github.jactor.persistence.service

import java.util.UUID
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import com.github.jactor.persistence.api.controller.toModel
import com.github.jactor.persistence.dto.UserModel
import com.github.jactor.persistence.entity.UserEntity
import com.github.jactor.persistence.person.PersonService
import com.github.jactor.persistence.repository.UserRepository
import com.github.jactor.shared.api.CreateUserCommand

@Service
class UserService(
    private val personService: PersonService,
    private val userRepository: UserRepository
) {
    fun find(username: String): UserModel? {
        return userRepository.findByUsername(username)
            .map { it.toModel() }
            .orElse(null)
    }

    fun find(id: UUID): UserModel? {
        return userRepository.findById(id)
            .map { it.toModel() }
            .orElse(null)
    }

    @Transactional
    fun update(userModel: UserModel): UserModel? {
        val uuid = userModel.persistentModel.id ?: throw IllegalArgumentException("User must have an id!")
        return userRepository.findById(uuid)
            .map { it.update(userModel) }
            .map { it.toModel() }
            .orElse(null)
    }

    fun create(createUserCommand: CreateUserCommand): UserModel {
        val user = createNewFrom(createUserCommand)

        if (user.id == null) {
            user.id = UUID.randomUUID()
        }

        return userRepository.save(user).toModel()
    }

    private fun createNewFrom(createUserCommand: CreateUserCommand): UserEntity {
        val personModel = createUserCommand.toPersonDto().toModel()
        val personEntity = personService.createWhenNotExists(person = personModel)
        val userEntity = UserEntity(user = createUserCommand.toUserDto().toModel())

        userEntity.setPersonEntity(personEntity)

        return userEntity
    }

    fun findUsernames(userType: UserEntity.UserType): List<String> {
        return userRepository.findByUserTypeIn(listOf(userType))
            .map { it.username ?: "username of user with id '${it.id} is null!" }
    }

    fun isAlreadyPresent(username: String): Boolean {
        return userRepository.findByUsername(username).isPresent
    }
}
